In [12]:
import numpy as np
from scipy import stats
import scipy.stats
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest

### [실습] 정규분포의 확률계산  
어떤 자동차 1리터당 주행거리가 평균 12km, 표준편차 3km인 정규분포를 이룬다.  
1). 1리터를 가지고 9km 이상 15km 이하를 달릴 확률은 얼마인가요?  
2). 1리터를 가지고 15km 이상을 달릴 확률은 얼마인가요?

In [22]:
mu = 12
sigma = 3
x = 9
x1 = 15
# 누적확률값 계산
prob_9 = stats.norm.cdf(x, mu, sigma) 
prob_15 = stats.norm.cdf(x1, mu, sigma)

print("{0}km 이상 {1}km이하를 달릴 확률은 {2:.3f}에 해당함".format(x, x1, prob_15-prob_9))
print("{0}km 이상 달릴 확률은 {1:.3f}에 해당함".format(x1, 1-prob_15))

9km 이상 15km이하를 달릴 확률은 0.683에 해당함
15km 이상 달릴 확률은 0.159에 해당함


### 실습(포아송 분포)
한 상점에는 평균 1시간에 3명씩 포아송분포로 손님이 도착한다.
이 때 10시에서 11시 사이에 다음이 발생할 확률을 구하시오.

1. 2명의 손님이 도착
2. 1명 이하의 손님이 도착
3. 3명 이상의 손님이 도착

In [45]:
mu = 3
# 확률질량값
prob = stats.poisson.pmf(2, mu)
# 누적확률값
cdf_prob = stats.poisson.cdf(1, mu)
cdf_prob2 = stats.poisson.cdf(2, mu)

print("10시에서 11시 사이에 {0}명의 손님이 도착할 확률 : {1:.2f}".format(2, prob))
print("10시에서 11시 사이에 {0}명 이하의 손님이 도착할 확률 : {1:.2f}".format(1, cdf_prob))
print("10시에서 11시 사이에 {0}명 이상의 손님이 도착할 확률 : {1:.2f}".format(3, 1-cdf_prob2))

# 손님이 2명 올 확률은 0.224
# 손님이 1명 이하로 올 확률은 0.199
# 손님이 3명 이상 올 확률은 0.577

10시에서 11시 사이에 2명의 손님이 도착할 확률 : 0.22
10시에서 11시 사이에 1명 이하의 손님이 도착할 확률 : 0.20
10시에서 11시 사이에 3명 이상의 손님이 도착할 확률 : 0.58


### [실습] 정규성 검정  
변수 Mat_A, Mat_B에 대해 각각 정규성을 검정

In [28]:
data = pd.read_csv('2sample_t_test.csv')
data = data.dropna() # 결측값 제거

A = data['Mat_A']
B = data['Mat_B']

Mat_A의 검정통계량 : 0.9062638878822327, p-value는 0.25630074739456177
Mat_B의 검정통계량 : 0.9729756116867065, p-value는 0.9169799089431763


In [32]:
# 검정통계량, p-value 계산
statistic_a, p_a = stats.shapiro(A)
statistic_b, p_b = stats.shapiro(B)

print('Mat_A의 검정통계량 : {0}, p-value는 {1:.3f}'.format(statistic_a, p_a))
print('Mat_B의 검정통계량 : {0}, p-value는 {1:.3f}'.format(statistic_b, p_b))

Mat_A의 검정통계량 : 0.9062638878822327, p-value는 0.256
Mat_B의 검정통계량 : 0.9729756116867065, p-value는 0.917


해석 : Mat_A의 경우, p-value가 0.256으로 정규성을 충족한다고 볼 수 있습니다.  
       Mat_B 또한, p-value가 0.917로 정규성을 충족한다고 볼 수 있습니다.

### [실습] 모평균 추정(모표준편차 아는 경우]  
새로운 품종의 감자 20 Box를 랜덤추출하여 무게를 측정할 결과가 다음과 같음  
(단, 모표준편차는 3.8로 알려져 있음)  
이 자료로부터 신품종 배추의 평균 무게에 대한 95% 신뢰구간을 구하시오

In [43]:
# 직접 계산

data = [18, 18, 20, 21, 20, 23, 19, 18, 17, 21,
       22, 20, 20, 21, 20, 19, 19, 18, 17, 19]
sigma = 3.8
sum_d = (18 + 18 + 20 + 21 + 20 + 23 + 19 + 18 + 17 + 21 +22 + 20 + 20 + 21 + 20 + 19 + 19 + 18 + 17 + 19)/20
lower = sum_d - 1.96*(3.8/np.sqrt(20))
upper = sum_d + 1.96*(3.8/np.sqrt(20))

print("신뢰구간 : ({0}, {1})".format(lower.round(2), upper.round(2)))

신뢰구간 : (17.83, 21.17)


In [49]:
# 함수 활용
lower, upper = stats.norm.interval(0.95, loc = np.mean(data), scale = 3.8/np.sqrt(20))
print("신뢰구간 : ({0}, {1})".format(lower.round(2), upper.round(2)))

신뢰구간 : (17.83, 21.17)


감자 20 Box 평균 무게에 대한 95% 신뢰구간은 17.83에서 21.17 입니다.  
즉, (17.83, 21.17) 내에 실제 모평균이 있을 가능성이 95% 라고 할 수 있습니다.

### [실습] 모평균 추정
제품개발자는 새로 개발된 강종의 평균 인장강도를 70kg/mm(제곱)으로 기대하고 있음.  
이를 검증하기 위해 시험 생산되는 강종 중 임의로 20개의 코일을 대상으로 랜덤하게 시편을 채취하였음(data)  
data로부터 표본평균은 71.50kg/mm(제곱), 표본표준편차는 1.933kg/mm(제곱)임을 알았음.  
이 모평균값의 95% 신뢰구간을 추정하시오.

In [51]:
data = [73, 71, 74, 69, 70, 73, 70, 68, 75, 72,
       70, 72, 73, 70, 70, 72, 71, 70, 75, 72]
lower, upper = stats.t.interval(0.95, len(data)-1, loc = np.mean(data), scale = scipy.stats.sem(data))
print("신뢰구간 : ({0}, {1})".format(lower.round(2), upper.round(2)))

신뢰구간 : (70.6, 72.4)


모평균 추정 결과, 새로 개발된 강종의 평균 인장강도의 95% 신뢰구간은 (70.6, 72.4) 입니다.  
95% 신뢰수준에서는 제품개발자가 기대한 인장강도 70kg/mm(제곱)은 모평균으로 추정될 수 없다.   